### PR TITLE
Add parameters to script in httpfuzzer processors

### PR DESCRIPTION
**Pull request by denniskniep**
… with DynamicFieldsPanel

### CODE REVIEW

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/FuzzerHttpMessageScriptProcessorAdapterUIHandler.java [hunk 1]
```diff
@@ -36,226 +44,318 @@
...
-            public int compareTo(ScriptUIEntry other) {
-                if (other == null) {
-                    return 1;
-                }
-                return scriptName.compareTo(other.scriptName);
-            }
-
-        }
-    }
+		HttpFuzzerMessageProcessorUIHandler<FuzzerHttpMessageScriptProcessorAdapter, FuzzerHttpMessageScriptProcessorAdapterUI> {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/FuzzerHttpMessageScriptProcessorAdapterUIHandler.java]: Since 200+ lines are marked as changed here it's really hard to tell what's relevant.
Did you change tabs or line endings or something?
(The content on this line _appears_ to be the same as the first removed line.)

Formatting changes should be done in a separate PR. 
https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java [hunk 2]
```diff
@@ -33,12 +35,7 @@
  * @see HttpFuzzerTaskProcessorUtils
  */
 public interface HttpFuzzerProcessorScript {
-
-    /**
-     * The name used to identify the type of this script, for example, in configurations.
-     */
-    final static String TYPE_NAME = "httpfuzzerprocessor";
-
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java]: White space change should be reverted.

https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java [hunk 3]
```diff
@@ -57,4 +54,6 @@
      * @throws ScriptException if an error occurs while executing the script
      */
     boolean processResult(HttpFuzzerTaskProcessorUtils utils, HttpFuzzResult result) throws ScriptException;
+  
 }
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java]: These two as well.

https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java [hunk 4]
```diff
@@ -19,6 +19,8 @@
  */
 package org.zaproxy.zap.extension.fuzz.httpfuzzer.processors;
 
+import java.util.Map;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScript.java]: Unused?

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
+package org.zaproxy.zap.extension.fuzz.httpfuzzer.processors;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java]: Should have the standard license header w/ 2017

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	public final static String TYPE_NAME = "httpfuzzerprocessor";
+	private static final Logger log = Logger.getLogger(HttpFuzzerProcessorScriptProxy.class);
+	HttpFuzzerProcessorScript script;
+	HttpFuzzerProcessorScriptWithParameters scriptWithParameters;
+	
+	public static HttpFuzzerProcessorScriptProxy create(ScriptWrapper scriptWrapper) throws Exception{
+		HttpFuzzerProcessorScriptProxy proxy = new HttpFuzzerProcessorScriptProxy();
+		ExtensionScript extensionScript = Control.getSingleton().getExtensionLoader().getExtension(ExtensionScript.class);
+        if (extensionScript != null) {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java]: Seems to be a mix of tabs and spaces for indents throughout.

File: src/org/zaproxy/zap/extension/fuzz/resources/Messages.properties [hunk 7]
```diff
@@ -438,6 +438,7 @@ fuzz.httpfuzzer.processor.scriptProcessor.error.persistMessageForAlert = An erro
 fuzz.httpfuzzer.processor.scriptProcessor.panel.script.label = Script:
 fuzz.httpfuzzer.processor.scriptProcessor.panel.warnNoScript.message = No script selected, a script must be selected first.
 fuzz.httpfuzzer.processor.scriptProcessor.panel.warnNoScript.title = No Script Selected
+fuzz.httpfuzzer.processor.scriptProcessor.panel.warn.title = Incorrect input
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/resources/Messages.properties]: Should be Title Caps: https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#text-titles-menu-items-options-and-tool-tips

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptWithParameters.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ * 
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ * 
+ * Copyright 2015 The ZAP Development Team
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptWithParameters.java]: 2017

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptWithParameters.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+public interface HttpFuzzerProcessorScriptWithParameters {    
+    
+    void processMessage(HttpFuzzerTaskProcessorUtils utils, HttpMessage message, Map<String, String> paramsValues) throws ScriptException;
+
+    boolean processResult(HttpFuzzerTaskProcessorUtils utils, HttpFuzzResult result, Map<String, String> paramsValues) throws ScriptException;
+    
+    String[] getRequiredParamsNames();
+
+	String[] getOptionalParamsNames();
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptWithParameters.java]: Seems to be a mix of tabs and space in this file for indents.

**Comment by denniskniep** [comment]
Sorry for the delay, but I have currently no time. I will probably tackle your comments next week.

**Comment by kingthorin** [comment]
@denniskniep no worries, we understand how life gets busy. Thanks for letting us know!

**Comment by denniskniep** [comment]
I added a new commit to this pull request

**Comment by denniskniep** [comment]
Hi @kingthorin, did you notice the new commit? Hope it tackles your review comments.

File: src/org/zaproxy/zap/extension/fuzz/ScriptUIEntry.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        result = prime * result + ((scriptName == null) ? 0 : scriptName.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/ScriptUIEntry.java]: This should be the first if, otherwise the condition above can NPE

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/FuzzerHttpMessageScriptProcessorAdapterUIHandler.java [hunk 11]
```diff
@@ -133,22 +146,44 @@ public FuzzerHttpMessageScriptProcessorAdapterUI copy() {
             AbstractHttpFuzzerMessageProcessorUIPanel<FuzzerHttpMessageScriptProcessorAdapter, FuzzerHttpMessageScriptProcessorAdapterUI> {
 
         private static final String SCRIPT_FIELD_LABEL = Constant.messages.getString("fuzz.httpfuzzer.processor.scriptProcessor.panel.script.label");
-
-        private final JPanel fieldsPanel;
+        private static final Logger log = Logger.getLogger(FuzzerHttpMessageScriptProcessorAdapterUIPanel.class);
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/FuzzerHttpMessageScriptProcessorAdapterUIHandler.java]: Should be full caps as a constant `LOG`

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.fuzz.httpfuzzer.HttpFuzzResult;
+import org.zaproxy.zap.extension.fuzz.httpfuzzer.HttpFuzzerTaskProcessorUtils;
+import org.zaproxy.zap.extension.script.ExtensionScript;
+import org.zaproxy.zap.extension.script.ScriptWrapper;
+
+public class HttpFuzzerProcessorScriptProxy implements HttpFuzzerProcessorScriptWithParameters {
+
+    public final static String TYPE_NAME = "httpfuzzerprocessor";
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java]: This should be ordered the same as the next one `static final`

File: src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    private static final Logger log = Logger.getLogger(HttpFuzzerProcessorScriptProxy.class);
+    HttpFuzzerProcessorScript script;
+    HttpFuzzerProcessorScriptWithParameters scriptWithParameters;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java]: These two private?

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/fuzz/httpfuzzer/processors/HttpFuzzerProcessorScriptProxy.java]: log above should be full caps too

**Comment by denniskniep** [comment]
@kingthorin I integrated your feedback. And added a Null check in class `AbstractHttpFuzzerMessageProcessorUIPanel` line 208 due to possible NPE when there are no scripts of this specific type.

**Comment by kingthorin** [comment]
Missed including the file in the PR?

**Comment by denniskniep** [comment]
Oh no sorry for this I copied the wrong classname, it is the class `FuzzerHttpMessageScriptProcessorAdapterUIPanel` in File `FuzzerHttpMessageScriptProcessorAdapterUIHandler.java`

**Comment by denniskniep** [comment]
@kingthorin 
Is now everything ok?

**Comment by kingthorin** [comment]
Looks ok to me. I grabbed the code and the UI looks ok. 

Have you used it for something? I'm not 100% sure I see the use case or user story for this. I added the template script to a fuzzer run and set a value for each of the required params then added a `print(paramValues);` in `processMessage` and ended up with `<eval>:61 ReferenceError: "paramValues" is not defined` ...

**Comment by denniskniep** [comment]
@kingthorin may be you missed a `s`? -> `print(paramsValues);`

The use case for me was that I would like to give some parameters into the script [showDifferences.js](https://github.com/zaproxy/community-scripts/blob/master/httpfuzzerprocessor/showDifferences.js)
 i.e. `removeOtherStatesFromStateColumn=true/false`

You can already change this in the script directly but it is not so convenient. Furthermore it is more explicit from the developer to the user how he can adjust the behaviour of the script without looking into the code.

**Comment by kingthorin** [comment]
@denniskniep thanks, I probably did miss an `s`.

Looking at the [IRC logs](http://logs.glob.uno/?c=mozilla%23websectools&s=11+Aug+2017&e=14+Aug+2017#c36944) I see that you were looking for me the other day. Sorry I missed you, and thanks for following up. Yes, your example use case is clear.

**Comment by denniskniep** [comment]
@kingthorin any news?

**Comment by thc202** [comment]
Yeah, will do.

**Comment by kingthorin** [comment]
Sorry I didn't manage to get to this tonight. I'm off and Friday and will try again then.

**Comment by denniskniep** [comment]
@kingthorin any updates?

**Comment by kingthorin** [comment]
@denniskniep thanks for following up. Sometime a nudge is helpful 👍 

Yes I've been able to successfully use it, you were right about me misnaming the variable.

**Comment by kingthorin** [comment]
Outstanding: 
* Help update. 
* Script caching (to prevent repeated parsing when unchanged)

**Comment by kingthorin** [comment]
@thc202 / @psiinon should the fuzzer help move into the add-on, versus being in core-help?

https://github.com/zaproxy/zap-core-help/wiki/HelpAddonsFuzzHttpmessageprocessors

**Comment by thc202** [comment]
The help is in the add-on, but for add-ons bundled in the main release the help is published in `zap-core-help` wiki instead of `zap-extensions` wiki.

**Comment by kingthorin** [comment]
Oh ok, I'll get that done shortly.

**Comment by thc202** [comment]
Thanks!

The parameters of the script are now being cached.

**Comment by kingthorin** [comment]
Help has been updated to include the new details:
image

**Comment by thc202** [comment]
Thanks @denniskniep and @kingthorin!
